### PR TITLE
Add check for public key presence on sm2 signing

### PR DESCRIPTION
**Pull request by nhorman**
SM2 requires that the public EC_POINT be present in a key when signing. If its not there we crash on a NULL pointer.  Add a check to ensure that its present, and raise an error if its not

##### Checklist
- [x] tests are added or updated

### CODE REVIEW

**Review by InfoHunter — APPROVED** [review]
Nice. Is it worth adding a description in the doc to inform the users that omitting public key leads to a signing failure in SM2?

**Comment by nhorman** [comment]
Certainly can't hurt, standby

**Comment by nhorman** [comment]
updated to document the requirement in EVP_PKEY-SM2.pod

File: doc/man7/EVP_PKEY-SM2.pod [hunk 1]
```diff
@@ -38,6 +38,9 @@ Getter that returns the default digest name.
 B<SM2> signatures can be generated by using the 'DigestSign' series of APIs, for
 instance, EVP_DigestSignInit(), EVP_DigestSignUpdate() and EVP_DigestSignFinal().
 Ditto for the verification process by calling the 'DigestVerify' series of APIs.
+Note that the SM2 algorithm requires the presence of the public key for signatures,
+as such the B<OSSL_PKEY_PARAM_PUB_KEY> option must be set on any key used in signature
+generation
```
Comment (InfoHunter) [inline doc/man7/EVP_PKEY-SM2.pod:43]: Nit: period is missing.

**Review by InfoHunter — APPROVED** [review]
Re-approve

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master, 3.3 and 3.2 branches. It does not cherry-pick cleanly to 3.1,3.0. @nhorman Could you please open a PR against 3.1 branch?

**Comment by nhorman** [comment]
Done

**Pull request closed without merging**